### PR TITLE
Suggest Replacement Types for Invalid Placeholders

### PR DESCRIPTION
**Pull request by CodaFi**
First, relax the restrictions on placeholder types appearing in interface types. In the past, we would pretty much replace every instance of them with an error type, which would lose information when we came around to apply the contextual type to any solutions for associated inference sources like initializers in pattern bindings or the return types of functions. Preserve these types in interface types so they can be opened later.

Second, now that placeholder types are preserved, we can open them wherever they appear in positions where they are used as contextual types. Use the checks we already run in the primaries to ban placeholders in top-level positions. Only now, use the inferred type of any associated expressions or statements to present the user with a contextual replacement type.

For parameters in functions and enum cases, we use any default argument expressions to get the contextual type.

For functions, we use a traversal similar to the opaque result type finder to find the type of any return statements in the program and present those as options.

Resolves rdar://82837146

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci test

File: lib/Sema/TypeCheckDeclPrimary.cpp [hunk 1]
```diff
@@ -967,11 +967,67 @@ static void checkInheritedDefaultValueRestrictions(ParamDecl *PD) {
...
+    // Placeholder types are banned for all parameter decls. We try to use the
+    // freshly-checked default expression's contextual type to suggest a
+    // reasonable type to insert.
+    if (expr && !expr->getType()->hasError()) {
+      param->diagnose(diag::placeholder_type_not_allowed_in_parameter)
+          .highlight(param->getSourceRange());
+      TypeChecker::notePlaceholderReplacementTypes(
+          ifacety, expr->getType()->mapTypeOutOfContext());
+    } else {
+      param->diagnose(diag::placeholder_type_not_allowed);
```
Comment (Jumhyn) [inline lib/Sema/TypeCheckDeclPrimary.cpp:1028]: Is there a reason why we use a generic fallback message here? Seems like the same diagnostic we use above, just without the fix-it, would be appropriate here as well.

**Review by Jumhyn — COMMENTED** [review]
This is awesome, major diagnostic improvement! I have a couple comments about diagnostics and some thoughts on allowing top-level placeholders wrt the evolution process.

File: test/Sema/placeholder_type.swift [hunk 2]
```diff
@@ -1,16 +1,17 @@
 // RUN: %target-typecheck-verify-swift
 
-let x: _ = 0 // expected-error {{placeholders are not allowed as top-level types}}
+let x: _ = 0
```
[the remark this answers is not in the document]
Comment (xwu) [reply test/Sema/placeholder_type.swift:3]: Without opening up the question of top-level use in general, where _not having a type annotation at all_ would compile as here, it’s not a load-bearing “use,” and perhaps could be considered not a “use” at all.

File: lib/AST/ASTContext.cpp [hunk 3]
```diff
@@ -3594,7 +3594,6 @@ GenericFunctionType *GenericFunctionType::get(GenericSignature sig,
                                               Optional<ExtInfo> info) {
   assert(sig && "no generic signature for generic function type?!");
   assert(!result->hasTypeVariable());
-  assert(!result->hasPlaceholder());
```
Comment (xedin) [inline lib/AST/ASTContext.cpp:3597]: Could you add an assert that verifies that there are placeholder no placeholder that originated from a type variable directly contain one (e.g. dependent member types with type variable base)?

File: include/swift/AST/DiagnosticsSema.def [hunk 4]
```diff
@@ -3683,6 +3681,12 @@ NOTE(descriptive_generic_type_declared_here,none,
 
 ERROR(placeholder_type_not_allowed,none,
       "type placeholder not allowed here", ())
+ERROR(placeholder_type_not_allowed_in_return_type,none,
+      "type placeholder may not appear in function return type", ())
+ERROR(placeholder_type_not_allowed_in_parameter,none,
+      "type placeholder may not appear in top-level parameter", ())
```
[the remark this answers is not in the document]
Comment (xedin) [reply include/swift/AST/DiagnosticsSema.def:3687]: I agree, we need to find something more user-friendly here.

File: include/swift/AST/DiagnosticsSema.def [hunk 5]
```diff
@@ -3683,6 +3681,12 @@ NOTE(descriptive_generic_type_declared_here,none,
...
+NOTE(replace_placeholder_with_inferred_type,none,
+     "replace the placeholder with the correct type %0", (Type))
```
Comment (xedin) [inline include/swift/AST/DiagnosticsSema.def:3689]: Both `the` are extraneous here according to Diagnostics.md and I'm not sure whether we should use `correct` at all...

File: test/Sema/placeholder_type.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (Jumhyn) [reply test/Sema/placeholder_type.swift:3]: Perhaps this wasn't the best line to attach the comment to, but we have another instance below that veers into the more problematic territory that @xwu noted:

```
let _: [String: Int] = dictionary(ofType: _.self) // used to fail, compiles under this PR
```

IMO we should seek core team guidance (if you haven't already) before accepting a change that directly violates the accepted language of SE-0315 (even though I'd be in favor of the change, evolution concerns notwithstanding 🙂).

**Comment by CodaFi** [comment]
@swift-ci test

**Comment by CodaFi** [comment]
@swift-ci test

**Comment by CodaFi** [comment]
⛵

**Pull request merged**